packages/ai-jsx/src/core/constrained-output.tsx

```diff
@@ -0,0 +1,141 @@
+/**
+ * This module defines affordances for constraining the output of the model
+ * into specific formats, such as JSON, YAML, or Markdown.
+ */
+import * as LLMx from '../index.js';
+import { ChatCompletion, SystemMessage, AssistantMessage, UserMessage } from './completion';
+import { memo } from './memoize';
+import yaml from 'js-yaml';
+
+function isJsonString(str: string): boolean | string {
+  try {
+    JSON.parse(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+function isYamlString(str: string): boolean | string {
+  try {
+    // try to parse the string as YAML
+    yaml.load(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+/**
+ * ChatCompletion that constrains the output to be a valid object format (e.g. JSON/YAML).
+ *
+ *
+ * @param type_name name of the type, to be used in the prompt
+ * @param validator function that returns true if the output is valid, or a string with an error message if not
+ *     // TODO: what should the signature of this function be?
+ * @param retries if output fails validation, how many times should we retry?
+ * @param verbose returns intermediate output if true
+ * @returns a string that validates as the given type (TODO: should I return the parsed object instead?)
+ */
+async function* ObjectFormatChatCompletion(
+  {
+    retries = 2,
+    verbose = false,
+    validator,
+    type_name,
+    children,
+    ...props
+  }: {
+    validator: (output: string) => boolean | string;
+    type_name: string;
+    retries?: number;
+    verbose?: boolean;
+    children: LLMx.Node[];
+  },
+  { render }: LLMx.RenderContext
+) {
+  const childrenWithCompletion = memo(
+    <ChatCompletion {...props}>
+      {children}
+      <SystemMessage>
+        Your response must be a valid {type_name}. Do not include ```{type_name.toLowerCase()} ``` code blocks.
+      </SystemMessage>
+    </ChatCompletion>
+  );
+
+  if (verbose) yield childrenWithCompletion;
+
+  let output = await render(childrenWithCompletion);
+  let valiationResults = validator(output);
+  if (valiationResults === true) {
+    return output;
```

Comment:
@petersalas rather than returning a string, what would you think about returning the parsed value?

If one wrote:

```tsx
const character = await render(<JsonOutput><GenerateFantasyCharacter /><JsonOutput>)
```

It seems nice to have `character` be parsed for me already, and not need to wrap it in `JSON.parse`.

Will it mess up the render machinery to be passing arbitrary values through, rather than assuming the output of every render is a string or another well-known value?

Reply:
Currently everything (eventually) renders to strings, but partial rendering is the (not super ergonomic) exception.

`JsonOutput` could render to say `<Json value={value} />` that internally does a `JSON.stringify` if forced to render.

And then the caller could do `await render(<JsonOutput />, { stop: (e) => e.tag === Json })` which will return an array containing strings and/or `Json` elements (from which the result could be plucked from props).

At the moment I think `JSON.parse` is nicer than ☝️, but this is an absolutely a scenario on my mind. Right now the bias is towards strings so that everything can be trivially passed back _into_ the LLM but I can imagine us supporting other atomic types that can be coerced into strings (or not if the model supports e.g. images!).